### ISSUE TITLE
Allowed memory size of xxx bytes exhausted (tried to allocate xxx bytes)

### ISSUE THREAD

**Issue by inquisitive-stha** [thephpleague/csv#298]
My Code
----------

```
        $path = storage_path('app/files/').$request->filename;
        $csv = Reader::createFromPath($path, 'r');
        $csv->setHeaderOffset(0); 
        $csv_header = $csv->getHeader(); 
        $sample_data = $csv->fetchOne();
        $sample_data = array_values($sample_data);
        $filename = $request->filename;
```

Allowed memory size get exhausted. I can print out just before the line $csv->getHeader() but not after that line. I don't get why it is getting exhausted.  How Could I solve it?

Size of CSV file is around 150MB and the total number of rows is around 500001 rows.

Testing on windows on the local machine but will upload it on Linux server. At the moment, receiving same error on both platform.

I have tried setting delimeter as  $csv->setDelimiter(';');  but get same error.

Also tried with $csv->setDelimiter("\r");  and "\n" but received same error 

Currently, max_file_upload_size is 200MB

EDIT:

I have also tried to detect delimiter as

```
use function League\Csv\delimiter_detect;
use League\Csv\Reader;

$reader = Reader::createFromPath('/path/to/file.csv', 'r');

$result = delimiter_detect($reader, [' ', '|'], 10);
```

But, neither this detect function is running successfully. It too got the memory exhausted.

**Comment by nyamsprod** [comment]
@inquisitive-stha thanks for using the library.
How can you be certain this issue is from the library and not any other part of your code. I'm asking because I've just tried your code on a 7M+ records CSV and I had no memory issue at all.
Are you sure your issue does not come from your upload script ?

**Comment by inquisitive-stha** [comment]
Yeah, it was working fine on other small file. Was it because of delimiter issue? I get stucked here.

**Comment by nyamsprod** [comment]
Again delimiter has nothing to do with your issue. And like I said I'm working with a 7M+ records CSV and have no memory issue. the library uses stream internally to reduce memory usage. That's why I'm asking hw can you tell this memory usage is coming from the library ... try your script with a file already present on the server if it work .... and it should then it means that the issue is not related to the library but to how you upload your file.

**Comment by inquisitive-stha** [comment]
Ok, let me explain it in detail.
1) It is successfully uploading the file. However, it takes quite a lot of time to upload the file, so I am checking it by directly passing the uploaded file path.
2) Why I am thinking it is delimiter issue. Below is the code snippet:
```
        if (($handle = fopen($path, "r")) !== FALSE) {
            while (($data = fgetcsv($handle, 1000, ",")) !== FALSE) {
                $num = count($data);
                print_r($data);
                exit();
            }
            fclose($handle);
        }
```
While trying it with above code. It get exhausted as usual. But if I replace that comma delimiter **","** with **"\r"**. Now, it shows data as:

```
    [0] => name,address,xx,xxx,xxx,xxx,xxx
    [1] => shrestha, saroj, new road,xx,xx,xxx,xxx
    [2] => 544770,47,Park,WILLIAMS,WILLIAM,J,,"WILLIAMS, WILL
```
Also, if I tried with **\n** it shows data but all in 0 index. Now, let me know what might be the issue.
Note, There is also **,** in data itself. Example: Single name column has last_name, first_name
But, I also tired with setting 
```
$csv->setDelimiter("\r");
$delimiter = $csv->getDelimiter();
echo $delimiter;
echo "==";
exit();
```
But, it does not show "\r" neither it push === to next line. Now, what do you think what might be the issue. If you are there, I am ready to get help with screen sharing too.

**Comment by nyamsprod** [comment]
> It is successfully uploading the file. However, it takes quite a lot of time to upload the file, so I am checking it by directly passing the uploaded file path.

That's already raises some concern IMHO. You are trying to parse a file which is not complete. Which leads to unexpected behaviour.

**Comment by inquisitive-stha** [comment]
Yeah, it take some extra minutes on server but on localhost it uploads quite faster. But, am 200% sure on uploading. As only after uploading, it redirects to another controller and then this process get started there. Also, to make you more confident about this, I just tried by copying the original file directly to the location. And tried with that, but same error. :(

**Comment by nyamsprod** [comment]
You could try using `Reader::createFromFilObject`  instead and see the error still occurs if yes then it's not the lib. If no then it's the stream object which behave strangely

**Comment by inquisitive-stha** [comment]
Is that 
```
$csv = Reader::createFromFileObject(new \SplFileObject($path));
```
Showing same error but at different function

```
 protected function seekRow(int $offset)
    {
        $this->document->setFlags(SplFileObject::READ_CSV | SplFileObject::READ_AHEAD | SplFileObject::SKIP_EMPTY);
        $this->document->setCsvControl($this->delimiter, $this->enclosure, $this->escape);
        $this->document->rewind();
 
        //Workaround for SplFileObject::seek bug in PHP7.2+ see https://bugs.php.net/bug.php?id=75917
        if (PHP_VERSION_ID >= 70200 && !$this->document instanceof Stream) {
            while ($offset !== $this->document->key() && $this->document->valid()) {
                $this->document->next();
            }
 
            return $this->document->current();
        }
```
So, what should I do, How to check which delimiter is set at the time.

**Comment by nyamsprod** [comment]
You do understand that the delimiter is not the issue. The issue is with your file. You set whatever delimiter you want with the setter method. Once set the delimiter does not change. The question no becomes What is the size of each record? Do you have thousand of fields or just a few? But as stated in the beginning today's I have worked with a 7M+ CSV without any memory issue using your provided code

**Comment by nyamsprod** [comment]
The easiest solution is to try to parse your CSV data using native PHP functions `fopen` or `SplFileObject` which is what this library is using under the hood. If you can not succeed with those this library won't help either.

**Comment by inquisitive-stha** [comment]
@nyamsprod Did you check with the file I have provided? Does it run on yours or get exhausted?

**Comment by inquisitive-stha** [comment]
@nyamsprod I am not saying anything bad to you and your library, I really appreciate you guys's hard work. I also, tried with other file with more rows. They work fine. But not with this file. I really appreciate your time and effort on helping for me. If you can't help me I don't think any other could help me, that's the only reason I am bothering to you. Sorry, for bothering you a lot. But, could you show me a way how can I get rid from this :(

**Comment by nyamsprod** [comment]
@inquisitive-stha I simply don't know . If it was a bug in the library I would look into it but since it is not I can only give you advices which are use directly PHP functions and if they fail then try to look into the file to see what is wrong with it.

**Comment by inquisitive-stha** [comment]
@nyamsprod Ok I figured out why this is not working. Whenever I tried directly with php functions fopen with ,(comma) delimiter and unlimited length it keeps on getting exhausted but after adding auto_detect_line_endings to deal with Mac line endings I am able to deal with that.
```
ini_set('auto_detect_line_endings',TRUE);  //ADDEDD LINE
$handle = fopen('/path/to/file','r');
while ( ($data = fgetcsv($handle) ) !== FALSE ) {
//process
}
ini_set('auto_detect_line_endings',FALSE);
```
So, in case of phpleague how can I do this. Is there something I am missing.
